### ISSUE TITLE
Use improved Glob to match files by pattern

### ISSUE THREAD

**Issue by DavertMik** [Codegyre/Robo#233]
@gabor-udvari 

> The whole findFiles method is an exact copy from the FlattenDir task. There is already a discussion about adding glob support for all the tasks. When that happens the findFiles method can be replaced.

Yes, please add this glob https://github.com/webmozart/glob implementation. It looks lightweight and stable
Please update all tasks where you think it might be needed

**Comment by burzum** [comment]
@gabor-udvari any update on this?

**Comment by burzum** [comment]
I don't like that https://github.com/webmozart/glob is a singleton. So my suggestion would be to implement a class around it and provide it by a trait to the tasks that simply creates a class instance and allows access to the glob by $this->glob()->match(), ->iterator()...

**Comment by gabor-udvari** [comment]
How would the enduser be able to switch between native and webmozart/glob?

Should we add a new function for it:
```php
$this->taskMinify( 'web/assets/*.css' )
     ->glob('native')
     ->run()
```

Or move the source argument definition to a new src function similar to npm:
```php
$this->taskMinify()
     ->src( 'web/assets/*.css', 'native')
     ->run()
```

**Comment by burzum** [comment]
@gabor-udvari the "end user" would not deal with it or change it. Except he implements a new task but then he makes a choice and sticks to it. You use another glob lib for a good reason. Otherwise you can stick to phps default method. I don't see much sense in changing the glob implementation any way once you've decided to use one. The main reason to use another glob implementation is that webmozart's one is more powerful.

If you read my previous posting again keep an eye on what glob() returns. It's an object. Different glob libs *will* have different interfaces. The arg of the glob() method, which would be provided by a trait. Returns an object of a specific glob implementation which is defined by it's 1st arg.

**Comment by DavertMik** [comment]
I agree with @burzum , what the purpose of using native implementation if @webmozart created a better one ) I think we don't need switching between those two implementations, and provide some sort of general interface. I'd like to keep everything simple here.

**Comment by boedah** [comment]
Isn't this a duplicate of #216?

**Comment by burzum** [comment]
Yes it is. Which one should we close? I think the other one, this here already discusses the implementation.

**Comment by webmozart** [comment]
Hi all :) The class is static for the reasons you mentioned. Globbing is a very low-level functionality and I don't think it makes sense to allow users to switch glob implementations, as in practice nobody does that. If I can help in any way, just let me know.

**Comment by burzum** [comment]
@webmozart I agree with you but I know @DavertMik likes it flexible :) But I personally don't like to do lots of static calls either. So creating a trait that wraps the static interface in methods that could be overriden by the users of Robo could be a good solution if you *really* have to do so for some reason.

**Comment by webmozart** [comment]
My approach to this issue generally is to wait until someone *really* has that issue and think about a solution then. It's much easier to find a solution for an actual problem than for an imaginary one. :)

**Comment by boedah** [comment]
Agree. 
I also smell a little bit of YAGNI in this one. 

Plus, as a bonus, if people need glob support, they use the API of webmozart's glob and can use this knowledge for other projects, which would not be the case if Robo wrapped stuff around...

**Comment by dlancea** [comment]
User-wise, the differences I can see between [native glob](http://php.net/manual/en/function.glob.php) and [webmozart/glob](https://github.com/webmozart/glob) are:

- Native syntax supports character classes and negative character classes, using the syntax `[]` and `[^]`. Webmozart glob does not support square brackets at all.
- Native supports `?` to match a single character, whereas Webmozart glob does not.
- Webmozart glob supports `/**/` to matches zero or more directory names, but native does not.

The last bullet is especially useful for the Robo project, given the nature of task runners, but this does show some backwards incompatibility. 

Maybe offer two glob functions, `glob()`, which uses native and `antGlob()`, which uses Webmozart's ant-like glob?

**Comment by webmozart** [comment]
I just added support for `?` to [webmozart/glob](https://github.com/webmozart/glob) which will be included in the next release (3.2). I'll see how complicated character classes would be.

**Comment by webmozart** [comment]
Character classes `[abc]` and inverted character classes `[^abc]` are now supported too.

**Comment by webmozart** [comment]
A new version 3.3 is now available that supports everything you asked for. :)

**Comment by burzum** [comment]
@webmozart awesome! Thank you!

@DavertMik how would you like to add this to Robo? Let's expose the static method calls through public methods added via a trait to the classes that need it?

**Comment by gabor-udvari** [comment]
Made the [first commit](https://github.com/gabor-udvari/Robo/commit/eb1ef102940673e3821dfc490cb161886cd198d6) on my own branch, not PR ready yet. @burzum, please check if it fits your idea with the wrapper class.

**Comment by burzum** [comment]
@gabor-udvari thank you for your contribution. I've reviewed the commit and left three comments. The last word about if this is going to be merged has @DavertMik.

**Comment by DavertMik** [comment]
I think that we are going to stick to webmozrt's implementation. Send a PR for it.

**Comment by greg-1-anderson** [comment]
Nice to have; not required for 1.0.
